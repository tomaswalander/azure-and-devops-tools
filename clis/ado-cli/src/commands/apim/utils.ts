import { OpenAPIV2 } from 'openapi-types';

type OpenApiV2Paths = OpenAPIV2.Document['paths'];

const methodsToCheck = [
  OpenAPIV2.HttpMethods.POST,
  OpenAPIV2.HttpMethods.GET,
  OpenAPIV2.HttpMethods.PATCH,
  OpenAPIV2.HttpMethods.PUT,
  OpenAPIV2.HttpMethods.DELETE,
  OpenAPIV2.HttpMethods.HEAD,
  OpenAPIV2.HttpMethods.OPTIONS,
];

const checkOperationIdOnMethod = (
  operationIds: string[],
  pathItemObject: OpenAPIV2.PathItemObject,
  method: OpenAPIV2.HttpMethods,
): OpenAPIV2.OperationObject | undefined => {
  const operationId =
    pathItemObject && pathItemObject[method]
      ? pathItemObject[method]?.operationId
      : undefined;
  if (operationId && operationIds.includes(operationId)) {
    return pathItemObject[method];
  }
  return undefined;
};

export const filterOpenApiSpecByOperationIds = (
  openApiSpec: OpenAPIV2.Document,
  operationIds: string[],
): OpenAPIV2.Document => {
  const filteredOpenApiSpec: OpenAPIV2.Document = {
    ...openApiSpec,
    paths: Object.keys(openApiSpec.paths).reduce(
      (previous: OpenApiV2Paths, path: string) => {
        const pathItemObject = openApiSpec.paths[path];
        const filteredPathItemObject: OpenAPIV2.PathItemObject = {};
        methodsToCheck.forEach(method => {
          const result = checkOperationIdOnMethod(
            operationIds,
            pathItemObject,
            method,
          );
          if (result) {
            filteredPathItemObject[method] = result;
          }
        });

        if (Object.keys(filteredPathItemObject).length > 0) {
          previous[path] = filteredPathItemObject;
        }

        return previous;
      },
      {} as OpenApiV2Paths,
    ),
  };

  return filteredOpenApiSpec;
};
