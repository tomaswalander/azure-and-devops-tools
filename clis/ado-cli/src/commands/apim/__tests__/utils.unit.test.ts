import { filterOpenApiSpecByOperationIds } from '../utils';

describe('utils', () => {
  describe('filterOpenApiSpecByOperationIds', () => {
    it('should return only the specified operation ids: ping and list-resources', () => {
      const result = filterOpenApiSpecByOperationIds(openApiSpecExample, [
        'ping',
        'list-resources',
      ]);
      expect(result.paths['/ping'].get).toBeDefined();
      expect(result.paths['/health']).toBeUndefined();
      expect(result.paths['/v1/resources'].post).toBeUndefined();
      expect(result.paths['/v1/resources'].get).toBeDefined();
      expect(result.paths['/v1/resources'].put).toBeUndefined();
      expect(result.paths['/v1/resources'].delete).toBeUndefined();
      expect(Object.keys(result.paths)).toHaveLength(2);
      expect(result).toMatchSnapshot();
    });
    it('should return only the specified operation ids: create-resource and list-resources', () => {
      const result = filterOpenApiSpecByOperationIds(openApiSpecExample, [
        'create-resource',
        'list-resources',
      ]);
      expect(result.paths['/ping']).toBeUndefined();
      expect(result.paths['/health']).toBeUndefined();
      expect(result.paths['/v1/resources'].post).toBeDefined();
      expect(result.paths['/v1/resources'].get).toBeDefined();
      expect(result.paths['/v1/resources'].put).toBeUndefined();
      expect(result.paths['/v1/resources'].delete).toBeUndefined();
      expect(Object.keys(result.paths)).toHaveLength(1);
      expect(result).toMatchSnapshot();
    });
    it('should return only the specified operation ids: create-, list-, update- and delete-resource', () => {
      const result = filterOpenApiSpecByOperationIds(openApiSpecExample, [
        'create-resource',
        'list-resources',
        'delete-resource',
        'update-resource',
      ]);
      expect(result.paths['/ping']).toBeUndefined();
      expect(result.paths['/health']).toBeUndefined();
      expect(result.paths['/v1/resources'].post).toBeDefined();
      expect(result.paths['/v1/resources'].get).toBeDefined();
      expect(result.paths['/v1/resources'].put).toBeDefined();
      expect(result.paths['/v1/resources'].delete).toBeDefined();
      expect(Object.keys(result.paths)).toHaveLength(1);
      expect(result).toMatchSnapshot();
    });
    it('should return only the specified operation ids: health, create-, update- and delete-resource', () => {
      const result = filterOpenApiSpecByOperationIds(openApiSpecExample, [
        'health',
        'create-resource',
        'delete-resource',
        'update-resource',
      ]);
      expect(result.paths['/ping']).toBeUndefined();
      expect(result.paths['/health'].get).toBeDefined();
      expect(result.paths['/v1/resources'].post).toBeDefined();
      expect(result.paths['/v1/resources'].get).toBeUndefined();
      expect(result.paths['/v1/resources'].put).toBeDefined();
      expect(result.paths['/v1/resources'].delete).toBeDefined();
      expect(Object.keys(result.paths)).toHaveLength(2);
      expect(result).toMatchSnapshot();
    });
  });
});

const openApiSpecExample = {
  swagger: '2.0',
  info: {
    title: 'Experimentations',
    description: 'This is the OpenAPI Document on Azure Functions',
    version: '1.0.0',
  },
  host: 'func-omnv-test-domain-experimentations.azurewebsites.net',
  basePath: '/api',
  schemes: ['https'],
  paths: {
    '/ping': {
      get: {
        tags: ['Monitoring'],
        operationId: 'ping',
        responses: {},
      },
    },
    '/health': {
      get: {
        tags: ['Monitoring'],
        operationId: 'health',
        responses: {},
      },
    },
    '/v1/resources': {
      get: {
        tags: ['Resources'],
        operationId: 'list-resources',
        responses: {},
      },
      post: {
        tags: ['Resources'],
        operationId: 'create-resource',
        responses: {},
      },
      put: {
        tags: ['Resources'],
        operationId: 'update-resource',
        responses: {},
      },
      delete: {
        tags: ['Resources'],
        operationId: 'delete-resource',
        responses: {},
      },
    },
  },
};
