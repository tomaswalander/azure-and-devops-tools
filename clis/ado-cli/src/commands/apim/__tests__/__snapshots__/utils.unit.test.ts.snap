// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils filterOpenApiSpecByOperationIds should return only the specified operation ids: create-, list-, update- and delete-resource 1`] = `
{
  "basePath": "/api",
  "host": "func-omnv-test-domain-experimentations.azurewebsites.net",
  "info": {
    "description": "This is the OpenAPI Document on Azure Functions",
    "title": "Experimentations",
    "version": "1.0.0",
  },
  "paths": {
    "/v1/resources": {
      "delete": {
        "operationId": "delete-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "get": {
        "operationId": "list-resources",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "post": {
        "operationId": "create-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "put": {
        "operationId": "update-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
    },
  },
  "schemes": [
    "https",
  ],
  "swagger": "2.0",
}
`;

exports[`utils filterOpenApiSpecByOperationIds should return only the specified operation ids: create-resource and list-resources 1`] = `
{
  "basePath": "/api",
  "host": "func-omnv-test-domain-experimentations.azurewebsites.net",
  "info": {
    "description": "This is the OpenAPI Document on Azure Functions",
    "title": "Experimentations",
    "version": "1.0.0",
  },
  "paths": {
    "/v1/resources": {
      "get": {
        "operationId": "list-resources",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "post": {
        "operationId": "create-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
    },
  },
  "schemes": [
    "https",
  ],
  "swagger": "2.0",
}
`;

exports[`utils filterOpenApiSpecByOperationIds should return only the specified operation ids: health, create-, update- and delete-resource 1`] = `
{
  "basePath": "/api",
  "host": "func-omnv-test-domain-experimentations.azurewebsites.net",
  "info": {
    "description": "This is the OpenAPI Document on Azure Functions",
    "title": "Experimentations",
    "version": "1.0.0",
  },
  "paths": {
    "/health": {
      "get": {
        "operationId": "health",
        "responses": {},
        "tags": [
          "Monitoring",
        ],
      },
    },
    "/v1/resources": {
      "delete": {
        "operationId": "delete-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "post": {
        "operationId": "create-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
      "put": {
        "operationId": "update-resource",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
    },
  },
  "schemes": [
    "https",
  ],
  "swagger": "2.0",
}
`;

exports[`utils filterOpenApiSpecByOperationIds should return only the specified operation ids: ping and list-resources 1`] = `
{
  "basePath": "/api",
  "host": "func-omnv-test-domain-experimentations.azurewebsites.net",
  "info": {
    "description": "This is the OpenAPI Document on Azure Functions",
    "title": "Experimentations",
    "version": "1.0.0",
  },
  "paths": {
    "/ping": {
      "get": {
        "operationId": "ping",
        "responses": {},
        "tags": [
          "Monitoring",
        ],
      },
    },
    "/v1/resources": {
      "get": {
        "operationId": "list-resources",
        "responses": {},
        "tags": [
          "Resources",
        ],
      },
    },
  },
  "schemes": [
    "https",
  ],
  "swagger": "2.0",
}
`;
